This is a patch to Wcl-2.7.  It does the following:

- Changes some of the defaults in Wc.tmpl so they are correct for
  redhat linux, including a fix to workaround some problems in the
  section that was not intended to be edited.  Non-linux machines
  should be unaffected, I used #ifdefs.

- Adds a routine to the library to allow a user-supplied substitution
  for XrmGetFileDatabase().  I needed this for the PERL module, so I can
  use PERL strings, instead of files.

- Fixes a couple minor #include bugs.

- Some very minor changes to support X11R6.

- Initial attempt at Motif 2.0 support.

To apply the patch:

1. Unpack the Wcl-2.7 sources somewhere and cd into that directory.

2. patch -s <this_file

Joe Buehler
jhpb@sarto.gaithersburg.md.us

*** ./Wc.tmpl	Tue Nov  1 13:21:28 1994
--- ./Wc.tmpl	Fri Aug  8 21:02:50 1997
***************
*** 34,48 ****
--- 34,59 ----
  ** 40 = X11R4 where XtNameToWidget barfs on Gadgets (early patch levels)
  ** 41 = X11R4 with fixed XtNameToWidget
  ** 50 = X11R5
+ ** 60 = X11R6
  */
+ 
+ #ifdef LinuxArchitecture
+ #define XRelease 60
+ #else
  #define XRelease 50
+ #endif
  
  /******************************************************************************
  ** Which flavors of libraries do you need?  Shared library systems by default
  ** do not build normal libraries, so you must ask for them if you need them.
  */
+ #ifdef LinuxArchitecture
+ #define ForceNormalLibs NO
+ #define MakeDebugLibs   NO
+ #else
  #define ForceNormalLibs YES
  #define MakeDebugLibs   YES
+ #endif
  #define MakeProfileLibs NO
  #define HasSharedLibraries YES
  
***************
*** 99,105 ****
  ** dlopen() and dlsym() and the macro RTLD_LAZY then you can use some
  ** wizzy dynamic linking stuff.  Unfortunately, gcc does not support this.
  */
! #if defined(SVR4) || defined(SunArchitecture) || defined(OSF1) && !HasGcc && !MakeProfileLibs
  #if HasSharedLibraries
  #define Has_dlopen_And_dlsym YES
  #endif
--- 110,116 ----
  ** dlopen() and dlsym() and the macro RTLD_LAZY then you can use some
  ** wizzy dynamic linking stuff.  Unfortunately, gcc does not support this.
  */
! #if defined(SVR4) || defined(SunArchitecture) || defined(OSF1) && !HasGcc && !MakeProfileLibs || defined(LinuxArchitecture)
  #if HasSharedLibraries
  #define Has_dlopen_And_dlsym YES
  #endif
***************
*** 277,283 ****
  /******************************************************************************
  ** gcc shared libraries do NOT seem to work.
  */
! #if HasGcc
  #undef HasSharedLibraries
  #define HasSharedLibraries NO
  #undef Has_dlopen_And_dlsym
--- 288,294 ----
  /******************************************************************************
  ** gcc shared libraries do NOT seem to work.
  */
! #if HasGcc && !defined(LinuxArchitecture)
  #undef HasSharedLibraries
  #define HasSharedLibraries NO
  #undef Has_dlopen_And_dlsym
***************
*** 366,368 ****
--- 377,384 ----
      LINTFLAGS = $(LINTOPTS) -DLINT $(ALLDEFINES)
  #endif
  
+ #ifdef LinuxArchitecture
+ /* missing in XFree86 imake setup */
+ #define _SharedLibMkdir()
+ #define _SharedCleanDir()
+ #endif
*** ./Wc/WcCreate.h	Thu Jul 21 13:49:10 1994
--- ./Wc/WcCreate.h	Fri Aug  8 20:59:35 1997
***************
*** 230,235 ****
--- 230,255 ----
  */
  typedef Widget (*WcWidgetConstructor) _((Widget,char*,ArgList,Cardinal));
  
+ /* -- Function pointer type used in overriding the default behavior
+  * for resource file loading.
+  */
+ typedef XrmDatabase (*WcDatabaseFunction) _((char *filename));
+ 
+ /* -- Use this function to override the default behavior for resource
+  * file loading.  Pass it a function pointer of the appropriate type
+  * (defined just above), pointing to the function that you have
+  * written to load resource files.  When Wcl needs to load a resource
+  * file, it will call your function, passing it the name of the file
+  * to be loaded.  Your function should return an XrmDatabase.
+  *
+  * Wcl uses the XrmGetFileDatabase() function if you do not set your
+  * own behavior.  Note that Wcl just wants a resource database back,
+  * it does not care whether you read a file to get it.
+  */
+ extern WcDatabaseFunction WcSetDatabaseFunction _((
+ 		WcDatabaseFunction func /* pointer to function to fetch resources */
+ ));
+ 
  /*====================================================================
   * Widget class, constructor, and callback proc registration routines:
   * Registrations are kept separate for each application context.
*** ./Wc/WcLoadRes.c	Fri Aug 27 18:10:57 1993
--- ./Wc/WcLoadRes.c	Fri Aug  8 21:11:05 1997
***************
*** 53,58 ****
--- 53,94 ----
  static void WcxLoadWarnTooLong _(( Widget, char* name, char* path, char* file));
  static void WcxLoadWarnMalloc _(( Widget, char* name ));
  
+ /* Override the Default Resource Fetching Function
+ *******************************************************************************
+     Argument is a pointer to a function that is used instead of the
+     Wcl default, when fetching resource files.
+     The previous saved function pointer is returned.
+ */
+ 
+ static WcDatabaseFunction WcGetFileDatabaseFunction;
+ 
+ WcDatabaseFunction WcSetDatabaseFunction( func )
+     WcDatabaseFunction func;
+ {
+ 	WcDatabaseFunction old_value = WcGetFileDatabaseFunction;
+ 	WcGetFileDatabaseFunction = func;
+ 	return(old_value);
+ }
+ 
+ /* Make Resource Database from a File
+ *******************************************************************************
+     Argument is name of file to build database from.
+     Return value is X resource database created from the file.
+     The default behavior is to use XrmGetFileDatabase(), but this can
+     be overridden by calling WcSetDatabaseFunction() with a pointer to
+ 	a function that the user has written himself.
+ */
+ 
+ static XrmDatabase WcGetFileDatabase( filename )
+     char *filename;
+ {
+ 	if (WcGetFileDatabaseFunction) {
+ 		return((*WcGetFileDatabaseFunction)(filename));
+ 	} else {
+ 		return(XrmGetFileDatabase(filename));
+ 	}
+ }
+ 
  /* Find Home Directory
  *******************************************************************************
      Used by WcLoadResourceFile and WcUserDefined.
***************
*** 245,251 ****
      char*        name;
      XrmDatabase* dbPointer;
  {
!     XrmDatabase rdb = XrmGetFileDatabase( name );
  
      if ( (XrmDatabase)0 != rdb )
      {
--- 281,287 ----
      char*        name;
      XrmDatabase* dbPointer;
  {
!     XrmDatabase rdb = WcGetFileDatabase( name );
  
      if ( (XrmDatabase)0 != rdb )
      {
***************
*** 297,303 ****
      WcStrCpy( path, homeDir );
      WcStrCat( path, from );
  
!     if ( (XrmDatabase)0 != (rdb = XrmGetFileDatabase( path )) )
      {
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( path );
--- 333,339 ----
      WcStrCpy( path, homeDir );
      WcStrCat( path, from );
  
!     if ( (XrmDatabase)0 != (rdb = WcGetFileDatabase( path )) )
      {
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( path );
***************
*** 326,332 ****
  					NULL, 0, NULL );
      if ( WcNonNull( filename ) )
      {
! 	XrmDatabase rdb = XrmGetFileDatabase( filename );
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( filename );
  	XtFree( filename );
--- 362,368 ----
  					NULL, 0, NULL );
      if ( WcNonNull( filename ) )
      {
! 	XrmDatabase rdb = WcGetFileDatabase( filename );
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( filename );
  	XtFree( filename );
***************
*** 360,366 ****
      WcStrCat( filename, "/" );
      WcStrCat( filename, name );
  
!     if ( (XrmDatabase)0 != (rdb = XrmGetFileDatabase(filename)) )
      {
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( filename );
--- 396,402 ----
      WcStrCat( filename, "/" );
      WcStrCat( filename, name );
  
!     if ( (XrmDatabase)0 != (rdb = WcGetFileDatabase(filename)) )
      {
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( filename );
***************
*** 460,466 ****
  
      if ( WcNonNull(filename) )
      {
! 	XrmDatabase rdb = XrmGetFileDatabase( filename );
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( filename );
  	XtFree( filename );
--- 496,502 ----
  
      if ( WcNonNull(filename) )
      {
! 	XrmDatabase rdb = WcGetFileDatabase( filename );
  	XrmMergeDatabases( rdb, dbPointer );
  	WcxLoadTrace( filename );
  	XtFree( filename );
***************
*** 492,498 ****
  	WcStrCpy( filename, xApplResDir );
  	WcStrCat( filename, name );
  
! 	if ( (XrmDatabase)0 != ( rdb = XrmGetFileDatabase( filename ) ) )
  	{
  	    XrmMergeDatabases( rdb, dbPointer );
  	    WcxLoadTrace( filename );
--- 528,534 ----
  	WcStrCpy( filename, xApplResDir );
  	WcStrCat( filename, name );
  
! 	if ( (XrmDatabase)0 != ( rdb = WcGetFileDatabase( filename ) ) )
  	{
  	    XrmMergeDatabases( rdb, dbPointer );
  	    WcxLoadTrace( filename );
***************
*** 513,519 ****
      WcStrCat( filename, "/" );
      WcStrCat( filename, name );
  
!     if ( (XrmDatabase)0 != ( rdb = XrmGetFileDatabase( filename ) ) )
  
      {
  	XrmMergeDatabases( rdb, dbPointer );
--- 549,555 ----
      WcStrCat( filename, "/" );
      WcStrCat( filename, name );
  
!     if ( (XrmDatabase)0 != ( rdb = WcGetFileDatabase( filename ) ) )
  
      {
  	XrmMergeDatabases( rdb, dbPointer );
*** ./Wc/XtName.c	Wed Mar 18 05:03:59 1992
--- ./Wc/XtName.c	Fri Aug  8 20:20:44 1997
***************
*** 1,4 ****
! #include <X11/Wc/COPY_X>
  
  /*
  * SCCS_data: @(#) XtName.c 1.4 92/03/18 11:02:30
--- 1,4 ----
! #include <X11/Wc/COPY_X.h>
  
  /*
  * SCCS_data: @(#) XtName.c 1.4 92/03/18 11:02:30
*** ./Xmp/Table.c	Thu Nov  3 13:35:25 1994
--- ./Xmp/Table.c	Fri Aug  8 20:20:44 1997
***************
*** 1989,1997 ****
  #else
  	    _XmMoveObject( (RectObj)(loc->w), x, y );
  #endif
  
  #endif /* end of Motif 1.x code */
- 	}
      }
  }
  
--- 1989,1997 ----
  #else
  	    _XmMoveObject( (RectObj)(loc->w), x, y );
  #endif
+ 	}
  
  #endif /* end of Motif 1.x code */
      }
  }
  
*** ./Xmp/XmpMotif.h	Wed Jul 20 17:26:09 1994
--- ./Xmp/XmpMotif.h	Fri Aug  8 20:20:44 1997
***************
*** 23,28 ****
--- 23,33 ----
  #include <Xm/FileSB.h>
  #include <Xm/Form.h>
  #include <Xm/Frame.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/IconG.h>
+ #include <Xm/IconH.h>
+ #endif
  #include <Xm/Label.h>
  #include <Xm/LabelG.h>
  #include <Xm/List.h>
***************
*** 41,51 ****
--- 46,72 ----
  #include <Xm/SeparatoG.h>
  #include <Xm/Separator.h>
  #include <Xm/Text.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/CSText.h>
+ #endif
  #ifndef _OLD_MOTIF
  #include <Xm/TextF.h>
  #endif
  #include <Xm/ToggleB.h>
  #include <Xm/ToggleBG.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/TearOffBP.h>
+ #endif
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/ComboBox.h>
+ #include <Xm/Container.h>
+ #include <Xm/Notebook.h>
+ #include <Xm/SpinB.h>
+ #include <Xm/GrabShell.h>
+ #endif
  
  #include <X11/Xmp/Table.h>
  
*** ./Xmp/XmpMotifP.h	Wed Jul 20 17:26:15 1994
--- ./Xmp/XmpMotifP.h	Fri Aug  8 20:20:44 1997
***************
*** 28,33 ****
--- 28,37 ----
  #include <Xm/CascadeBGP.h>
  #include <Xm/PushBGP.h>
  #include <Xm/ToggleBGP.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/IconGP.h>
+ #endif
  
  /* XmPrimitive Subclasses
  */
***************
*** 36,41 ****
--- 40,49 ----
  #include <Xm/ScrollBarP.h>
  #include <Xm/SeparatorP.h>
  #include <Xm/TextP.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/CSTextP.h>
+ #endif
  
  /* TextField: Motif 1.0 doesn't have them, the TextFP.h collides with TextP.h,
   * and they are useless anyway.  Nevertheless, at least get TextF.h if there.
***************
*** 53,58 ****
--- 61,70 ----
  #include <Xm/DrawnBP.h>
  #include <Xm/PushBP.h>
  #include <Xm/ToggleBP.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/TearOffBP.h>
+ #endif
  
  /* XmManager Subclasses
  */
***************
*** 72,77 ****
--- 84,96 ----
  #include <Xm/FileSB.h>
  #endif
  #include <Xm/FileSBP.h>
+ #if (defined(XmVERSION) && XmVERSION > 1)
+ /* Motif 2.0 */
+ #include <Xm/ComboBoxP.h>
+ #include <Xm/ContainerP.h>
+ #include <Xm/NotebookP.h>
+ #include <Xm/SpinBP.h>
+ #endif
  
  /* Shell Subclasses
  */
*** ./Xmp/XmpRegAll.c	Fri Jul 22 12:48:31 1994
--- ./Xmp/XmpRegAll.c	Fri Aug  8 20:20:44 1997
***************
*** 102,108 ****
   RCO( "XmCreateFormDialog",		XmCreateFormDialog		);
   RCO( "XmCreateFrame",			XmCreateFrame			);
   RCO( "XmCreateInformationDialog",	XmCreateInformationDialog	);
! #if defined(XmREVISION) && XmREVISION >= 2
   RCO( "XmCreateTemplateDialog",		XmCreateTemplateDialog		);
  #endif
   RCO( "XmCreateLabel",			XmCreateLabel			);
--- 102,108 ----
   RCO( "XmCreateFormDialog",		XmCreateFormDialog		);
   RCO( "XmCreateFrame",			XmCreateFrame			);
   RCO( "XmCreateInformationDialog",	XmCreateInformationDialog	);
! #if (defined(XmVERSION) && XmVERSION > 1) || (defined(XmREVISION) && XmREVISION >= 2)
   RCO( "XmCreateTemplateDialog",		XmCreateTemplateDialog		);
  #endif
   RCO( "XmCreateLabel",			XmCreateLabel			);
***************
*** 143,148 ****
--- 143,171 ----
   RCO( "XmCreateToggleButtonGadget",	XmCreateToggleButtonGadget	);
   RCO( "XmCreateWarningDialog",		XmCreateWarningDialog		);
   RCO( "XmCreateWorkingDialog",		XmCreateWorkingDialog		);
+ #if (defined(XmVERSION) && XmVERSION > 1)
+  /* Motif 2.0 */
+  RCO( "XmCreateComboBox",			XmCreateComboBox			);
+  RCO( "XmCreateCommandDialog",		XmCreateCommandDialog		);
+  RCO( "XmCreateContainer",			XmCreateContainer			);
+  RCO( "XmCreateCSText",				XmCreateCSText				);
+  RCO( "XmCreateDragIcon",			XmCreateDragIcon			);
+  RCO( "XmCreateDropDownComboBox",	XmCreateDropDownComboBox	);
+  RCO( "XmCreateDropDownList",		XmCreateDropDownList		);
+  RCO( "XmCreateGrabShell",			XmCreateGrabShell			);
+  RCO( "XmCreateIconGadget",			XmCreateIconGadget			);
+  RCO( "XmCreateIconHeader",			XmCreateIconHeader			);
+  RCO( "XmCreateNotebook",			XmCreateNotebook			);
+  RCO( "XmCreateScrolledCSText",		XmCreateScrolledCSText		);
+  RCO( "XmCreateSimpleCheckBox",		XmCreateSimpleCheckBox		);
+  RCO( "XmCreateSimpleMenuBar",		XmCreateSimpleMenuBar		);
+  RCO( "XmCreateSimpleOptionMenu",	XmCreateSimpleOptionMenu	);
+  RCO( "XmCreateSimplePopupMenu",	XmCreateSimplePopupMenu		);
+  RCO( "XmCreateSimplePulldownMenu",	XmCreateSimplePulldownMenu	);
+  RCO( "XmCreateSimpleRadioBox",		XmCreateSimpleRadioBox		);
+  RCO( "XmCreateSpinBox",			XmCreateSpinBox				);
+  RCO( "XmCreateTemplateDialog",		XmCreateTemplateDialog		);
+ #endif
  
  /* -- register Motif widget classes */
   RCP("xmArrowButtonWidgetClass",	xmArrowButtonWidgetClass	);
***************
*** 217,220 ****
--- 240,271 ----
   RCP("XmToggleButton",			xmToggleButtonWidgetClass	);
   RCP("xmToggleButtonGadgetClass",	xmToggleButtonGadgetClass	);
   RCP("XmToggleButtonGadget",		xmToggleButtonGadgetClass	);
+ #if (defined(XmVERSION) && XmVERSION > 1)
+  /* Motif 2.0 */
+ #if 0
+  RCP("xmClipWindowWidgetClass",		xmClipWindowWidgetClass		);
+  RCP("XmClipWindow",				xmClipWindowWidgetClass		);
+ #endif
+  RCP("xmComboBoxWidgetClass",		xmComboBoxWidgetClass		);
+  RCP("XmComboBox",					xmComboBoxWidgetClass		);
+  RCP("xmContainerWidgetClass",		xmContainerWidgetClass		);
+  RCP("XmContainer",					xmContainerWidgetClass		);
+  RCP("xmCSTextWidgetClass",			xmCSTextWidgetClass			);
+  RCP("XmCSText",					xmCSTextWidgetClass			);
+  RCP("xmDragOverShellWidgetClass",	xmDragOverShellWidgetClass	);
+  RCP("XmDragOverShell",				xmDragOverShellWidgetClass	);
+  RCP("xmGrabShellWidgetClass",		xmGrabShellWidgetClass		);
+  RCP("XmGrabShell",					xmGrabShellWidgetClass		);
+  RCP("xmIconGadgetClass",			xmIconGadgetClass			);
+  RCP("XmIconGadget",				xmIconGadgetClass			);
+  RCP("xmNotebookWidgetClass",		xmNotebookWidgetClass		);
+  RCP("XmNotebook",					xmNotebookWidgetClass		);
+  RCP("xmSpinBoxWidgetClass",		xmSpinBoxWidgetClass		);
+  RCP("XmSpinBox",					xmSpinBoxWidgetClass		);
+  RCP("xmTearOffButtonWidgetClass",	xmTearOffButtonWidgetClass	);
+  RCP("XmTearOffButton",				xmTearOffButtonWidgetClass	);
+ #if 0
+  RCP("XmVendorShellWidgetClass",	XmVendorShellWidgetClass	);
+ #endif
+ #endif
  }
*** ./Xp/XpAthena.h	Tue Nov  1 17:29:25 1994
--- ./Xp/XpAthena.h	Fri Aug  8 20:20:44 1997
***************
*** 42,48 ****
  #include <X11/Xaw/Scrollbar.h>
  #include <X11/Xaw/StripChart.h>
  #include <X11/Xaw/Text.h>
! #ifdef R5
  #include <X11/Xaw/Panner.h>
  #endif /*R5*/
  
--- 42,48 ----
  #include <X11/Xaw/Scrollbar.h>
  #include <X11/Xaw/StripChart.h>
  #include <X11/Xaw/Text.h>
! #if (defined(R5) || defined(R6))
  #include <X11/Xaw/Panner.h>
  #endif /*R5*/
  
***************
*** 52,58 ****
  #include <X11/Xaw/Toggle.h>
  
  /* Command */
! #ifdef R5
  #include <X11/Xaw/Repeater.h>
  #endif
  
--- 52,58 ----
  #include <X11/Xaw/Toggle.h>
  
  /* Command */
! #if (defined(R5) || defined(R6))
  #include <X11/Xaw/Repeater.h>
  #endif
  
***************
*** 74,80 ****
  #include <X11/Xaw/Box.h>
  #include <X11/Xaw/Form.h>
  #include <X11/Xaw/Paned.h>
! #ifdef R5
  #include <X11/Xaw/Porthole.h>
  #include <X11/Xaw/Tree.h>
  #endif /*R5*/
--- 74,80 ----
  #include <X11/Xaw/Box.h>
  #include <X11/Xaw/Form.h>
  #include <X11/Xaw/Paned.h>
! #if (defined(R5) || defined(R6))
  #include <X11/Xaw/Porthole.h>
  #include <X11/Xaw/Tree.h>
  #endif /*R5*/
***************
*** 85,89 ****
--- 85,90 ----
  #include <X11/Xaw/Viewport.h>
  
  #undef R5
+ #undef R6
  
  #endif /* _XpAthena_h_ */
